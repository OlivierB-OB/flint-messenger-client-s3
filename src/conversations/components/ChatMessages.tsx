import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import React, { Fragment, useRef, useEffect } from 'react';
import { IConversationMessage } from '../types';
import { ChatMessage } from './ChatMessage';

export interface IChatMessagesProps {
  conversationId?: string;
  messages: IConversationMessage[];
  conversationSeen: (id: string) => void;
}

export function ChatMessages({ messages, conversationId, conversationSeen }: IChatMessagesProps) {
  const ref = useRef<HTMLElement>(null);
  useEffect(() => {
    if (conversationId) conversationSeen(conversationId);
  }, [conversationId, messages, conversationSeen]);
  useEffect(() => ref.current?.scrollIntoView(false));
  if (!messages.length) return <span>There is currently no message in the conversation...</span>;
  return (
    <Fragment>
      <List>
        {messages.map((message) => (
          <ListItem key={message._id}>
            <ChatMessage message={message} />
          </ListItem>
        ))}
      </List>
      <span ref={ref} />
    </Fragment>
  );
}
