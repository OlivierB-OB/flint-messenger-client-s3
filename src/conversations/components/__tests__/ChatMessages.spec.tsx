import React from 'react';
import { shallow, mount } from 'enzyme';
import { ChatMessages } from '../ChatMessages';

describe('ChatMessages', () => {
  it('should display a feedback if there is no messages in the conversation', async () => {
    const props: any = {
      conversationId: '123',
      messages: [],
      conversationSeen: jest.fn(),
    };
    const component = shallow(<ChatMessages {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should display the conversation messages', async () => {
    const props: any = {
      conversationId: '123',
      messages: [
        { _id: '123', content: 'message 1' },
        { _id: '456', content: 'message 2' },
      ],
      conversationSeen: jest.fn(),
    };
    const component = shallow(<ChatMessages {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should call the callback conversationSeen on mount', async () => {
    const props: any = {
      conversationId: '123',
      messages: [],
      conversationSeen: jest.fn(),
    };
    mount(<ChatMessages {...props} />);
    expect(props.conversationSeen).toHaveBeenCalledWith('123');
  });

  it('should not call the callback conversationSeen if conversation if is unavailable', async () => {
    const props: any = {
      messages: [],
      conversationSeen: jest.fn(),
    };
    mount(<ChatMessages {...props} />);
    expect(props.conversationSeen).not.toHaveBeenCalled();
  });
});
