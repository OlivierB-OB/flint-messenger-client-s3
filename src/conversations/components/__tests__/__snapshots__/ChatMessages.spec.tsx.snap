// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessages should display a feedback if there is no messages in the conversation 1`] = `
<span>
  There is currently no message in the conversation...
</span>
`;

exports[`ChatMessages should display the conversation messages 1`] = `
<Fragment>
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))
      key="123"
    >
      <ChatMessage
        message={
          Object {
            "_id": "123",
            "content": "message 1",
          }
        }
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      key="456"
    >
      <ChatMessage
        message={
          Object {
            "_id": "456",
            "content": "message 2",
          }
        }
      />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
  <span />
</Fragment>
`;
